src/Game/AI.cpp

```diff
@@ -3,12 +3,234 @@
+            scoreDiff = env->team2->score - env->team1->score;
+        }
+        if (seeker->isFined) return val;
+        if(scoreDiff < -30){
+            if(seeker->position == env->snitch->position){
+                val = -1000;
+            }
+            else{
+                val = 200 / gameController::getDistance(seeker->position, env->snitch->position);
+            }
+        }
+        else {
+            val = 1000 / gameController::getDistance(seeker->position, env->snitch->position) + 1;
+        }
+
+        return val;
+    }
+
+    double evalKeeper(const std::shared_ptr<gameModel::Keeper> keeper,
+                      const std::shared_ptr<gameModel::Environment> env) {
+        double val = 0;
+        int scoreDiff = 0;
+        if(env->team1->hasMember(keeper)){
+            scoreDiff = env->team1->score - env->team2->score;
+        }
+        else{
+            scoreDiff = env->team2->score - env->team1->score;
+        }
+        if (keeper->isFined) return val;
+        //If keeper has quaffle
+        if (keeper->position == env->quaffle->position) {
+            val += 100;
+            if (env->isPlayerInOwnRestrictedZone(keeper)) {
+                if (scoreDiff > -40) val += 20;
+                if (scoreDiff >= 40) val += 500;
+            } else {
+                if(scoreDiff < -30){
+                    val += 100 + getHighestGoalRate(env, keeper) * 10;
+                }
+                else if(scoreDiff > 30){
+                    val += 200 + getHighestGoalRate(env, keeper);
+                }
+                else{
+                    val += 100 + getHighestGoalRate(env, keeper) * 2;
+                }
+            }
+        } else {
+
+        }
+
+        return val;
+    }
+
+    double evalChaser(const std::shared_ptr<gameModel::Chaser> chaser,
+                      const std::shared_ptr<gameModel::Environment> env) {
+        double val = 0;
+
+        if (chaser->isFined) return val;
+
+        int scoreDiff = 0;
+        if(env->team1->hasMember(chaser)){
+            scoreDiff = env->team1->score - env->team2->score;
+        }
+        else{
+            scoreDiff = env->team2->score - env->team1->score;
+        }
+        //If Chaser holds quaffle
+        if (chaser->position == env->quaffle->position) {
+            if(scoreDiff < -30){
+                val += 100 + getHighestGoalRate(env, chaser) * 10;
+            }
+            else if(scoreDiff > 30){
+                val += 200 + getHighestGoalRate(env, chaser);
+            }
+            else{
+                val += 100 + getHighestGoalRate(env, chaser) * 2;
+            }
+        }
+        else{
+            val += 100 / gameController::getDistance(chaser->position, env->quaffle->position);
+        }
+        return val;
+    }
+
+    double evalBludgers(const std::shared_ptr<gameModel::Environment> env) {
+        double val = 0;
+
+        std::array<std::shared_ptr<gameModel::Bludger>, 2> bludgers = env->bludgers;
```

Comment:
arbeite doch direkt auf dem env. Wenn du sicherstellen willst, dass die Bludger usw nicht verändert werden, dann leg dir ne lokale Copie an mit
```suggestion
        std::array<const std::shared_ptr<const gameModel::Bludger>, 2> &bludgers = env->bludgers;
```